default/python/AI/InvasionAI.py

```diff
@@ -315,7 +316,14 @@ def evaluate_invasion_planet(planet_id, secure_fleet_missions, verbose=True):
 
     system_id = planet.systemID
 
-    # make sure the target planet is not stealthed
+    # by using the following instead of simply relying on stealth meter reading, can (sometimes) plan ahead even if
+    # planet is temporarily shrouded by an ion storm
+    if not EspionageAI.colony_detectable_by_empire(planet_id, empire_id=fo.empireID(), default_result=False):
+        debug("InvasionAI predicts planet id %d to be stealthed" % planet_id)
+        return [0, 0]
+
+    # make sure the target planet was not extra-stealthed somehow its system was last viewed
+    # this test below augments the tests above, but by itself it can be thrown off by temporary combat-related sighting
```

Comment:
This looks incorrect to me. We may "predict" a planet to be stealthed eventhough we currently have perfect vision of it (because we did not consider some special/building/tech...). Is it intended to abort in that case?

How does this interact with the partial visibility checks below? My understanding is that if the planet is stealthed to us (but has been seen at some point), then we have basic visibility of it. If it not stealthed to us and it is in detection range, then we have *partial* visibility.

So, in an ion storm, we would only have *basic* visibility while possibly having partial visibility of the system (?). In that case, even if we predict that we can see the planet, the checks below do not care about that.

Or am I misremembering stuff regarding the stealth meters?

Reply:
> We may "predict" a planet to be stealthed eventhough we currently have perfect vision of it (because we did not consider some special/building/tech...). Is it intended to abort in that case?

It's my intention to be diligent about augmenting the code to keep the assessment as accurate as possible, in which case the far most likely reason (perhaps only reason if we are successful with the diligence) that a planet could still be visible even if predicted to be stealthed would be that it was temporarily exposed due to a Derelict Scout.  And so it was my intention to let it err on the side of aborting if it predicts no visibility. 

The alternative is that we add an override check for current-turn partial visibility.  If the partial visibility was due to a temporary effect then then hopefully the AI would correct course soon enough to avoid committing/wasting too many resources on the false opportunity.  Consider though even the case that the unexpected visibility is because of a persistent effect not yet considered by the function (such as, currently it does not consider Lighthouses)-- if the lighthouse is owned by the enemy, and we have no permanent bases in detection range and the detection range is being provided by a scout that might not remain in the area, then it could well be that in a couple turns we would lose the actual visibility, fall back to predicting no visibility, and then abort the in-progress mission that fleets had begun travel for, or  expect to have no use for something being built.  

I suppose, though, that those costs of sometimes building something which then loses its original purpose but can be repurposed, or temporarily pursuing a mission  which then has to be dropped, are probably less than the cost of ignoring a target that could in fact be had.  Let's mull on it just a bit more, but after talking this out I suppose I am leaning towards letter current actual visibility override a prediction of no visibility, as you suggest.

> My understanding is that if the planet is stealthed to us (but has been seen at some point), then we have basic visibility of it. If it not stealthed to us and it is in detection range, then we have partial visibility.

The last sentence is correct.  Adding a "in detection range" to the middle sentence makes it essentially correct, although the "(but has been seen at some point)" doesn't really matter and there are also some other ways that basic visibility can be granted.

> How does this interact with the partial visibility checks below? ... So, in an ion storm, we would only have basic visibility while possibly having partial visibility of the system (?). In that case, even if we predict that we can see the planet, the checks below do not care about that.  Or am I misremembering stuff regarding the stealth meters?

I had double checked that issue a few days ago when I first added these adjustments, but doing another double check now I saw a code section I had overlooked before, so in the case that we own a planet or ship in that same system you are right.  Especially now that we have the preceding prediction check, it seems this comparison of planet to system stealth should just be taken out.

Reply:
I've changed this section to simply create a debug logging statement but not actually reject the target planet.